### ISSUE TITLE
bsd_arandom.rs should loop on 256-byte chunks

### ISSUE THREAD

**Issue by niacat** [rust-random/getrandom#151]
The assumption that it does a short read for long buffers isn't always true. After reviewing the code in the NetBSD kernel we decided to adopt this behaviour moving forward (because it seems safer after reviewing real-world code using it), but older NetBSD releases will simply return E2BIG if more than 256 bytes are requested.

https://github.com/NetBSD/src/commit/7161bc211696634c0d0bc55c41a341bf919206e0

The semantics should be the same as getentropy(), which does loop on 256 byte chunks. The only difference is that the output size should be checked (however, the NetBSD kernel at least will never return short reads when <=256 bytes are requested, and will otherwise always return 256 bytes).

**Comment by newpavlov** [comment]
>The only difference is that the output size should be checked

Hm, does it mean we should use something like this?
```rust
for chunk in dest.chunks_mut(256) {
    sys_fill_exact(chunk, kern_arnd)?
}
```
It will mean that we will use two loops and an additional `EINTR` check.

**Comment by newpavlov** [comment]
@niacat 
Can you please create a PR with appropriate changes?

One option is to leave code as-is, since IIUC on NetBSD it should work without any issues.

**Comment by newpavlov** [comment]
@dhardy 
I think we can close this issue? It looks like adding additional chunking on our side will only make our code more complex with no apparent benefit.

**Comment by dhardy** [comment]
@newpavlov can I leave it to you or @josephlr?

**Comment by newpavlov** [comment]
Ok, I will wait for @josephlr to chime-in or for @niacat to argue why current code will not work properly.

**Comment by niacat** [comment]
The code will not work because the system call will error (return -1) if more than 256 bytes are requested. As mentioned previously, on NetbSD 9.0 and previous releases, the semantics of KERN_ARND are very similar to getentropy. I can't necessarily review or PR this since I don't know Rust, but there's clearly already code that deals with the 256 byte limit in your getentropy backend.

**Comment by josephlr** [comment]
@niacat thanks for the explanation here. I agree that we should make the change along the lines @newpavlov describes.

I think we were basing this on our understanding of FreeBSD code (which handles this issue fine). However, for both FreeBSD and NetBSD, only 256 bytes can be recieved from the kernel at a time, so this check doesn't (significantly) lose efficancy.

**Comment by newpavlov** [comment]
Ah, sorry, I misunderstood you. Your point is that old NetBSDs will not work with current code, while I was looking only at the new version of the kernel code.

**Comment by josephlr** [comment]
#156 is up, it pretty much does the obvious thing here (looping over 256 byte chunks)

**Comment by josephlr** [comment]
Closed as #156 is merged
